### PR TITLE
Converting PillButtonBar to vNext

### PR DESCRIPTION
**Pull request by edjamesmsft**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Tokenized PillButtonBar, moving PillButtonBar into the vNext controls. As some of these states are not covered by the figma definitions, best-effort was done on color choices. The following bug is open on the design team to track this: https://ghe-us.microsoft.com/Microsoft-Design-System-Studio/Fluent-Mobile/issues/77

### Verification

Manually tested in light and dark mode, different overriding color themes, different dimensions and with localization in mind.

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| image
 | image |

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/814)

### CODE REVIEW

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 1]
```diff
@@ -123,12 +115,13 @@ open class PillButtonBar: UIScrollView {
         }
     }
 
-    private var stackView: UIStackView = {
+    private var stackView = UIStackView()
```
Comment (huwilkes) [inline ios/FluentUI/Pill Button Bar/PillButtonBar.swift:118]: What's the reason for the change from stackView: UIStackView = {}() to stackView = UIStackView() with a different init function? Were the tokens not initialized yet, or something?

File: ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        updateForCurrentTheme()
+    }
+
+    @objc open func didChangeAppearanceProxy() {
+        updateForCurrentTheme()
+    }
+
+    override func updateForCurrentTheme() {
+        let currentTheme = theme
+        let appearanceProxy = currentTheme.MSFPillButtonBarTokens
```
Comment (huwilkes) [inline ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:33]: You probably don't need currentTheme if you're only using it to set up appearanceProxy.

File: tools/sgen/input/FluentUIStyle.yml [hunk 3]
```diff
@@ -403,6 +403,57 @@ AP_MSFPersonaViewTokens:
...
+  minButtonsSpacing: $Spacing.xSmall
+  minButtonVisibleWidth: $Spacing.large
+  minButtonWidth: 56pt
+  minHeight: 28pt
+  sideInset: $Spacing.medium
+  hoverBackgroundColor: [ primary: "FluentUIColor(light: NamedColor(grey94), dark: NamedColor(grey8))",
+                          onBrand: "FluentUIColor(light: $Colors.Brand.shade10, dark: NamedColor(grey8))" ]
+
+AP_MSFPillButtonTokens:
+  bottomInset: 6pt
```
Comment (rdeassis) [inline tools/sgen/input/FluentUIStyle.yml:418]: nit: can you sort the tokens alphabetically?

**Review by rdeassis — CHANGES_REQUESTED** [review]
Can you update the description according to the following?

1. Update the title with a more descriptive title.
2. Remove the space in between the square brackets so that the iOS checkbox renders correctly.
3. Remove the line breaks between the "|" characters and the images so the table renders correctly.
4. Add the vnext-prototype label to the PR.

File: ios/FluentUI.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -534,6 +542,10 @@
 		497DC2D62418585D008D86F8 /* Pill Button Bar */ = {
 			isa = PBXGroup;
 			children = (
+				D6F4098E274F1A1C001B80D4 /* PillButtonBarTokens.swift */,
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI.xcodeproj/project.pbxproj:545]: Oh, interesting, I didn't realize these were manually ordered. Will change that shortly.

File: ios/FluentUI/Pill Button Bar/PillButton.swift [hunk 5]
```diff
@@ -182,49 +186,46 @@ open class PillButton: UIButton {
...
                 }
             }
         }
     }
 
+    private var pillButtonTokens: MSFPillButtonTokens
+
     private struct Constants {
-        static let bottomInset: CGFloat = 6.0
         static let font = UIFont.systemFont(ofSize: 16, weight: .regular)
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:229]: There were no tokens available of the correct type. I will add this to the bug I filed on the design team.

File: ios/FluentUI/Pill Button Bar/PillButton.swift [hunk 6]
```diff
@@ -147,12 +151,12 @@ open class PillButton: UIButton {
...
 
-    private let unreadDotLayer: CALayer = {
-        let unreadDotLayer = CALayer()
-        unreadDotLayer.bounds.size = CGSize(width: Constants.unreadDotSize, height: Constants.unreadDotSize)
-        unreadDotLayer.cornerRadius = Constants.unreadDotSize / 2
-        return unreadDotLayer
-    }()
+    private var unreadDotLayer = CALayer()
+
+    private func initUnreadDotLayer() {
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:156]: Thanks for the suggestion! I was hoping someone would have a better way to do this :)

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 7]
```diff
@@ -358,32 +357,32 @@ open class PillButtonBar: UIScrollView {
...
     private func adjustButtonsForCurrentScrollFrame() {
-        var visibleWidth = frame.width - (Constants.minButtonsSpacing + Constants.minButtonVisibleWidth)
-        var visibleButtonsWidth = Constants.sideInset
+        var visibleWidth = frame.width - (pillButtonBarTokens.minButtonsSpacing + pillButtonBarTokens.minButtonVisibleWidth)
+        var visibleButtonsWidth = pillButtonBarTokens.sideInset
         var visibleButtonCount = 0
         for button in buttons {
             button.layoutIfNeeded()
-            visibleButtonsWidth += button.frame.width
+            visibleButtonsWidth! += button.frame.width
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:365]: I will think on the logic that makes the most sense here. I hesitate to default to a value of 16 because I generally prefer not to use magic numbers (especially since the purpose of this change is to move the definition of all of these attributes/values/tokens into the token system, rather than hard coded).

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 8]
```diff
@@ -600,7 +601,7 @@ extension PillButtonBar: UIPointerInteractionDelegate {
         if let window = window, customPillButtonBackgroundColor == nil, index < buttons.count {
             let pillButton = buttons[index]
             if !pillButton.isSelected {
-                pillButton.customBackgroundColor = PillButton.hoverBackgroundColor(for: window, for: pillButton.style)
+                pillButton.customBackgroundColor = (pillButtonStyle == .primary ? pillButtonBarTokens.hoverBackgroundColor : pillButtonBarTokens.onBrandHoverBackgroundColor)
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:604]: I learned a new thing to test! I'll look into this.

File: ios/FluentUI.xcodeproj/project.pbxproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.xcodeproj/project.pbxproj:545]: Yep Xcode is all manual. You can right-click the editor in Xcode and have it alphabetize for you though!

File: ios/FluentUI/Pill Button Bar/PillButton.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:229]: Looks like the spec calls for Body2, which maps to `"Font(15px, regular)"`. This is a bit different than the current, non-vNext control, but it does seem like we could switch over.

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:365]: Actually the problem seems to be that `pillButtonBarTokens.sideInset` is declared as a `CGFloat!`, which confuses the auto-typer sometimes. If you explicitly name the tyupe of `visibleButtonsWidth` on line 361 like so:

```swift
var visibleButtonsWidth: CGFloat = pillButtonBarTokens.sideInset
```

then all is well, and you can remove all the force unwraps (`!`) from here on out.

File: tools/sgen/input/FluentUIStyle.yml [hunk 9]
```diff
@@ -403,6 +403,57 @@ AP_MSFPersonaViewTokens:
...
+  horizontalInset: $Spacing.medium
+  topInset: 6pt
+  unreadDotOffsetX: 6pt
+  unreadDotOffsetY: 3pt
+  unreadDotSize: 6pt
+  backgroundColor: [ rest: "FluentUIColor(light: NamedColor(grey94), dark: NamedColor(grey8))",
```
Comment (mischreiber) [inline tools/sgen/input/FluentUIStyle.yml:424]: I know this is a spec issue, but it's still frustrating that so many of these colors are `NamedColor` values (e.g. Global Tokens).

File: ios/FluentUI/Pill Button Bar/PillButton.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:231]: Where are you seeing Body2? I just did a sync with the upstream branch and don't see Body2 anywhere in the FluentUIStyle doc.

File: tools/sgen/input/FluentUIStyle.yml (hunk 3 above)
Comment (edjamesmsft) [reply tools/sgen/input/FluentUIStyle.yml:418]: I was keeping logically related tokens close together (colors in one section, formatting/shape related tokens in another). Would you prefer a purely alphabetized list, or a middle ground, where they are logically separate but alphabetized within those sections?

File: ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    override init() {
+        super.init()
+
+        self.themeAware = true
+
+        updateForCurrentTheme()
+    }
+
+    @objc open func didChangeAppearanceProxy() {
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:27]: How can I test to verify whether this is or isn't happening?

[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:27]: Also note, CCB is UIKit, not SwiftUI, and does not have the themeDidUpdate/notification for updating tokens. So either we both have the same problem, or it isn't a problem.

[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:27]: And one last thing - what is this coding paradigm called? I'd like to do a bit more research, but I'm not certain what to search for :)

[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:27]: > Also note, CCB is UIKit, not SwiftUI, and does not have the themeDidUpdate/notification for updating tokens. So either we both have the same problem, or it isn't a problem.

FWIW, I'm fairly certain from my testing that CCB **does** have this same issue :) As they stand right now, it appears that our other UIKit vNext controls (at least the one's I've looked at) will only update to a new theme _after_ they are dismissed from their window and re-added. This isn't something that we would have noticed in our test app, since we always remove and re-add to change themes (by moving around the navigation stack).

@edjamesmsft, feel free to duplicate the logic used by the `DrawerController` and its `MSFDrawerTokens` that @rdeassis mentioned. That's also a UIKit control, and appears to be correct in its updating.

> And one last thing - what is this coding paradigm called? I'd like to do a bit more research, but I'm not certain what to search for :)

Not sure if it has a strict name, but it's basically a callback. Kinda like a delegation pattern-type thing, but not the "strict" `Delegate` pattern that UIKit often uses.

File: ios/FluentUI/Pill Button Bar/PillButton.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:230]: Ah yes, you're right that it's not currently called out as a discrete token.

In this case, I followed a trail of breadcrumbs:
- First I looked at the Figma spec for this component and saw that it asked for `Body2`.
- When I looked in `FluentUIStyle.yml`, I also saw that `Body2` wasn't listed. Sadly this isn't surprising, as that `yml` file is a bit out of date (hence my token work 😺)
- Instead, I looked back at the Figma spec to see how `Body2` is defined, and saw that it's a 15px "regular" font.
- I went back to `FluentUIStyle.yml` and saw that several other components refer to `"Font(15px, regular)"` directly as their desired font value.

So for now, I think we can follow the pattern set by `Avatar` and `Button` tokens and just directly set `"Font(15px, regular)"` as the desired token value. This'll be nicely cleaned up once the NEW new token system supports fonts!

[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:230]: Thanks for the walkthrough, super helpful! I've done so and it looks good so far :)

File: tools/sgen/input/FluentUIStyle.yml (hunk 9 above)
Comment (edjamesmsft) [reply tools/sgen/input/FluentUIStyle.yml:425]: I agree, but I don't have much choice in the matter right now :(

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:596]: I believe this is related to the previously-discussed "theme not updating" bug. Resolving for now, this will be fixed in a separate change.

File: ios/FluentUI/Pill Button Bar/PillButton.swift [hunk 11]
```diff
@@ -147,12 +158,14 @@ open class PillButton: UIButton {
         }
     }
 
-    private let unreadDotLayer: CALayer = {
-        let unreadDotLayer = CALayer()
-        unreadDotLayer.bounds.size = CGSize(width: Constants.unreadDotSize, height: Constants.unreadDotSize)
-        unreadDotLayer.cornerRadius = Constants.unreadDotSize / 2
-        return unreadDotLayer
-    }()
+    private lazy var unreadDotLayer: CALayer = {
```
Comment (rdeassis) [inline ios/FluentUI/Pill Button Bar/PillButton.swift:161]: I know we spoke about this lazy instantiation of this CALayer in a previous iteration, but I just though of a scenario that should be considered:

Token update scenario:
Because its initialization reads from tokens that can be changed, this property needs to be recreated to reflect the tokens change.

And that also applies to the stackview property below in the bar.

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 12]
```diff
@@ -592,7 +593,7 @@ extension PillButtonBar: UIPointerInteractionDelegate {
         if let window = window, customPillButtonBackgroundColor == nil, index < buttons.count {
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:593]: Ah, I think I asked about this specific warning in another conversation. The reason I didn't address it was because I thought it was pre-existing, as I hadn't changed that code directly, and the discussion landed on "fix that in another change to keep your current PR as focused on just this feature as possible." Looking at the diff now I realize it's a new warning because I removed the code that directly references the window. I will fix this shortly!

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:596]: https://office.visualstudio.com/APEX/_workitems/edit/5652140 for the task to track fixing this more correctly

File: ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI/Pill Button Bar/PillButtonBarTokens.swift:27]: https://office.visualstudio.com/APEX/_workitems/edit/5652140 for the task to fix this more correctly

File: ios/FluentUI/Pill Button Bar/PillButton.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButton.swift:164]: This new version is not quite what @rdeassis had in mind, I think. Speaking for them for a second (sorry @rdeassis if I get it wrong!), this version will always create and initialize the unread dot's `CALayer`, even if we never use it. Since that's likely to be the case for most users, it'd be nice to have it be both initialized and set up only when it's being used, which Swift allows for with the `lazy` keyword.

But we can't just tag the variable as `lazy`, since it's unconditionally initialized during the button's `init` method (which invokes `initUnreadDotLayer`). The solution @rdeassis suggested would only execute the lambda to create and initialize `unreadDotLayer` the first time it's accessed, which we can delay until when it's being requested for display.

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift (hunk 1 above)
Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:118]: Reopening for the same comments as on the `unreadDotLayer`, where we can just initialize this lazily and it _should_ work with the new token lookup.

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 13]
```diff
@@ -115,13 +115,15 @@ open class PillButtonBar: UIScrollView, FluentUIWindowProvider {
         }
     }
 
-    private var stackView = UIStackView()
+    private lazy var stackView: UIStackView = {
+        initStackView()
+    }()
```
Comment (mischreiber) [inline ios/FluentUI/Pill Button Bar/PillButtonBar.swift:120]: Looks good! And I think it can be simplified even further. The closure here simply invokes this method when run, so it can be optimized away like so:

```swift
private lazy var stackView: UIStackView = initStackView()
```

Comment (mischreiber) [reply ios/FluentUI/Pill Button Bar/PillButtonBar.swift:120]: Same with `unreadDotLayer` above.

File: ios/FluentUI/Pill Button Bar/PillButtonBar.swift [hunk 14]
```diff
@@ -542,12 +541,14 @@ open class PillButtonBar: UIScrollView {
     }
 
     private func updateHeightConstraint() {
-        var maxHeight = Constants.minHeight
-        buttons.forEach { maxHeight = max(maxHeight, $0.frame.size.height) }
+        var maxHeight = pillButtonBarTokens.minHeight
+        buttons.forEach { maxHeight = max(maxHeight!, $0.frame.size.height) }
         if let heightConstraint = heightConstraint, maxHeight != heightConstraint.constant {
-            heightConstraint.constant = maxHeight
+            heightConstraint.constant = maxHeight!
```
Comment (rdeassis) [inline ios/FluentUI/Pill Button Bar/PillButtonBar.swift:547]: Let's not use explicitly unwrap (`maxHeight!` operator).
You can either do guard let or provide a fallback (default) value if the value is not available from the tokens object.

File: tools/sgen/output/MSFNotificationTokens.generated.swift [hunk 15]
```diff
@@ -150,7 +150,7 @@ extension FluentUIStyle {
 
 		// MARK: - outlineColor 
 		open var outlineColor: UIColor {
-			return UIColor(named: "FluentColors/clear", in: FluentUIFramework.resourceBundle, compatibleWith: nil)!
+			return UIColor(named: "FluentColors/clear", in: FluentUIFramework.colorsBundle, compatibleWith: nil)!
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply tools/sgen/output/MSFNotificationTokens.generated.swift:153]: This was an unintentional change - all I did was run sgen after making my isolated token changes, but I didn't make any changes to MSFNotificationTokens in the yaml. I figured it was better for everything to be aligned (as in, maybe someone forgot to check in MSFNotificationTokens.generated.swift in the sgen output and this is fixing that mistake) than for this to continue to be an issue. I can revert if preferred.

Is there another reason this change would have happened?

[the remark this answers is not in the document]
Comment (mischreiber) [reply tools/sgen/output/MSFNotificationTokens.generated.swift:153]: This is a side-effect of two old conflicting PRs intersecting. The first was mine:

The first was @jeaniehuynh's:
- #715
This created tokens for NotificationView using the classic single-bundle approach for colors and images. While this was out in PR, I submitted a separate change:

- #748
This PR separated our resource bundle into two separate bundles, `resourceBundle` for iOS-specific resources, and `colorsBundle` for our cross-platform colors (it was necessary for Swift Package Manager stuff).

The latter never converted the tokens for the former, so it's just been waiting for someone to re-run sgen and get the new values.

The code is already correct in `vnext-prototype`, so there's no harm in this fix. Just a side-effect of the fact that this is the first time anyone's re-run sgen since those two PRs both made it to `vnext-prototype`.
